frontend/lib/resolution/resolution-queries.cpp

```diff
@@ -2209,223 +2284,282 @@ resolveFunctionByPoisQuery(Context* context,
+  } else if (visitor.initResolver) {
+    if (auto newSignature = visitor.initResolver->finalize()) {
+      if (newSignature != sig) outNewSignature = newSignature;
+    }
+  }
+
+  // then, resolve '=' and add any copy init/deinit calls as needed
+  callInitDeinit(visitor);
+
+  // then, handle return intent overloads and maybe-const formals
+  adjustReturnIntentOverloadsAndMaybeConstRefs(visitor);
 
-    // then, handle return intent overloads and maybe-const formals
-    adjustReturnIntentOverloadsAndMaybeConstRefs(visitor);
+  // check that throws are handled or forwarded
+  // TODO: Call for initializers as well, and remove checks in the resolver.
+  if (!isInitializer) checkThrows(context, rr, fn);
 
-    // check that throws are handled or forwarded
-    checkThrows(context, resolutionById, fn);
+  // TODO: can this be encapsulated in a method?
+  resolvedPoiInfo.swap(visitor.poiInfo);
+  resolvedPoiInfo.setResolved(true);
+  resolvedPoiInfo.setPoiScope(nullptr);
 
-    // TODO: can this be encapsulated in a method?
-    resolvedPoiInfo.swap(visitor.poiInfo);
-    resolvedPoiInfo.setResolved(true);
-    resolvedPoiInfo.setPoiScope(nullptr);
+  auto ret = toOwned(new ResolvedFunction(sig, fn->returnIntent(),
+                                          std::move(rr),
+                                          std::move(resolvedPoiInfo),
+                                          std::move(visitor.returnType)));
+  return ret;
+}
 
-    owned<ResolvedFunction> resolved
-        = toOwned(new ResolvedFunction(sig, fn->returnIntent(),
-                  std::move(resolutionById),
-                  resolvedPoiInfo,
-                  visitor.returnType));
+static const ResolvedFunction* const&
+resolveFunctionByInfoQuery(Context* context,
+                           const TypedFnSignature* sig,
+                           PoiInfo poiInfo) {
+  QUERY_BEGIN(resolveFunctionByInfoQuery, context, sig, poiInfo);
 
-    // Store the result in the query under the POIs used.
-    // If there was already a value for this revision, this
-    // call will not update it. (If it did, that could lead to
-    // memory errors).
+  const TypedFnSignature* finalSignature = sig;
+  CallerDetails empty;
+
+  auto resolved = resolveFunctionByInfoImpl(context, finalSignature, sig,
+                                            std::move(poiInfo),
+                                            empty);
+
+  // Make a copy of the resolved 'PoiInfo' for use later.
+  auto rPoiInfo = resolved->poiInfo();
+
+  // Finalizing an initializer may produce a new, different signature as the
+  // output. In this case we copy the results into a second 'ResolvedFunction'
+  // that maps to the newly produced signature.
+  // TODO: Do we need this copy, or is there a way that we can avoid it?
+  if (sig->isInitializer() && finalSignature != sig) {
+    auto r = resolved.get();
+    auto resolvedCopy = toOwned(new ResolvedFunction(finalSignature,
```

Comment:
I think it could be, but I think my rationale was I wanted to try and have the `...Impl` be a pure computation and avoid setting any queries if possible.